### PR TITLE
Don't require entityset when writing responses

### PR DESCRIPTION
**Pull request by mikepizzo**
### Issues

Enables writing responses without specifying an entityset.

### Description

UnknownEntitySet is treated as containment -- path becomes path through parent
Id of entities in unknown entity set is path if known, otherwise null
ContextUrl is path, if known, otherwise type

### Checklist (Uncheck if it is not completed)

- [x ] *Test cases added*
- [x ] *Build and test with one-click build and test script passed*

### Additional Work
-Might be able to get rid of ThrowIfMissingTypeInfo (or validate that it relates only to type info, not entity set info)

### CODE REVIEW

File: src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs [hunk 1]
```diff
@@ -393,6 +393,12 @@ internal override string GetOperationTitle(string operationName)
         /// </returns>
         internal override bool TryGetIdForSerialization(out Uri id)
         {
+            if(this.ResourceMetadataContext.TypeContext.NavigationSourceName == null)
+            {
+                id = null;
+                return false;
+            }
+
             id = this.ResourceMetadataContext.Resource.IsTransient ? null : this.GetId();
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs:402]: Id could be 'null' at Line 402, but it always returns 'true'? Now, the new return could be 'false'? what's the difference?

File: src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs [hunk 2]
```diff
@@ -131,11 +131,11 @@ private void ComputeAndCacheId()
                 case EdmNavigationSourceKind.Singleton:
                     uri = this.ComputeIdForSingleton();
                     break;
+                // Treat UnknownEntitySet as a containment
+                case EdmNavigationSourceKind.UnknownEntitySet:
                 case EdmNavigationSourceKind.ContainedEntitySet:
                     uri = this.ComputeIdForContainment();
                     break;
-                case EdmNavigationSourceKind.UnknownEntitySet:
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs:137]: Is it a behavior breaking change?

File: src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs [hunk 3]
```diff
@@ -152,7 +152,7 @@ private void ComputeAndCacheId()
         /// </returns>
         private Uri ComputeId()
         {
-            if (this.ResourceMetadataContext.KeyProperties.Any())
+            if (this.ResourceMetadataContext.KeyProperties.Any() && this.ResourceMetadataContext.TypeContext.NavigationSourceName != null)
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs:155]: move the  'null' testing ahead of 'Any()' could be better perf?

File: src/Microsoft.OData.Core/ODataResourceTypeContext.cs [hunk 4]
```diff
@@ -55,23 +55,23 @@ private ODataResourceTypeContext(IEdmStructuredType expectedResourceType, bool t
         /// </summary>
         public virtual string NavigationSourceName
         {
-            get { return this.ValidateAndReturn(default(string)); }
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/ODataResourceTypeContext.cs:58]: can we just use default property get?

public virtual string NavigationSourceName { get; }

**Review by xuzhg — COMMENTED** [review]
I left some comments and looking forward the PR

File: src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs (hunk 1 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs:402]: For a transient resource (like in aggregation) null *is* the right value for the id. In this case, the id isn't null, it just can't be computed. 

We have existing code that checks the boolean result of the TryGetIdForSerialization and handles it differently than true with a null value.

File: src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs (hunk 2 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs:137]: It is turning an error case into a success case.

File: src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs (hunk 3 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/Evaluation/ODataConventionalIdMetadataBuilder.cs:155]: :thumbsup: 

I like it!

File: src/Microsoft.OData.Core/ODataResourceTypeContext.cs (hunk 4 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/ODataResourceTypeContext.cs:58]: :thumbsup: 

yes; that would be simpler. good catch.

**Review by xuzhg — APPROVED** [review]

**Pull request merged**